pkg/causality/internal/slots.go

```diff
@@ -0,0 +1,90 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"container/list"
+	"sync"
+)
+
+// Eq describes objects that can be compared for equality.
+type Eq[T any] interface {
+	Equals(other T) bool
+}
+
+// Slots implements slot-based conflict detection.
+// It holds references to E, which can be used to build
+// a DAG of dependency.
+type Slots[E Eq[E]] struct {
+	mu    sync.Mutex
+	slots map[int64]*list.List
+
+	numSlots int64
+}
+
+// NewSlots creates a new Slots.
+func NewSlots[E Eq[E]](numSlots int64) *Slots[E] {
+	return &Slots[E]{
+		slots:    make(map[int64]*list.List),
+		numSlots: numSlots,
+	}
+}
+
+// Add adds an elem to the slots and calls onConflict for each case
+// where elem is conflicting with an existing element.
+// Note that onConflict can be called multiple times with the same
+// dependee.
+func (s *Slots[E]) Add(elem E, keys []int64, onConflict func(dependee E)) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for _, key := range keys {
+		needInsert := true
+		if elemList, ok := s.slots[key%s.numSlots]; ok {
+			if elemList.Len() > 0 {
+				lastElem := elemList.Back().Value.(E)
+				if lastElem.Equals(elem) {
+					needInsert = false
+				} else {
+					onConflict(lastElem)
+				}
+			}
+		} else {
+			s.slots[key%s.numSlots] = list.New()
+		}
+
+		if needInsert {
+			s.slots[key%s.numSlots].PushBack(elem)
```

Comment:
I have come up with an idea.
Since we use `slot` to compress the key space and construct dependency DAG therefore, we may suffer from some `false positive` dependency([a]->[a',b]->[a,c]) and long path depenpency link([a]->[a,b]->[a,c]).

May be we can `increasing path compression` in `Union-find` algorithm to find find more accurate the dependency DAG and shorten the path, so we can resolve the node faster.

Reply:
I'm confused by your notation for paths. Can you clarify what `DAG([a]->[[a,b],[a,c]]) and shorten the path([a]->[[a,b],[a,c]])` means?

Reply:
`DAG([a]->[[a,b],[a,c]])` need to be corrected as `DAG([a]->[[a',b],[a,c]])`.
It means, for a single slot, key 1(%3), 4(%3), 7(%3) may be assigned to the this slot, so we may have some `false positive`  dependency like [7, x] -> [4,x] -> [1, x] -> [7',x]. 
Every time we resolve the dependency, it need to follow the dependency chain, like [7',x] first,  [1, x] second, and then [4,x], [7, x] for last.

But the [7,x] is depent on [7',x] actually, not [4,x]. So if we can correct the dependency chain increasingly, it may be helpful. For example:
1. [7, x] -> [4,x] -> [1, x] -> [7',x]
2. [4,x] -> [1, x] -> [7',x]
               |
     [7,x]--
3. [4,x] -> [1, x] -> [7',x]
                               |
                     [7,x]--
 As we find the right dependency item, after we resolve [7',x], we can resolve [7,x] next.
 Just a shot in my mind.